keps/sig-node/2535-ensure-secret-pulled-images/README.md

```diff
@@ -165,57 +201,340 @@ ensure all tenants have rights to the images that are already present on a host.
+for credential reverification (and therefore ignored for caching).
+
+The rules that define what should be ignored are set by the `imagePullCredentialsVerificationPolicy`:
+- `NeverVerify` matches the previous behavior from user perspective. Kubelet still creates
+   records of image pulls as these would be required should the policy change later.
+- `NeverVerifyPreloadedImages` images pulled without a secret and those pulled
+  outside the kubelet process are ignored (new default behavior)
+- `NeverVerifyAllowlistedImages` - images pulled without a secret and those pulled
+  outside the kubelet process but mentioned in `preloadedImagesVerificationAllowlist` are ignored
+- `AlwaysVerify` - all the image pulls need reverification.
+
+#### Writing to the Cache
+
+There are two different stages when the kubelet updates its cache:
+1. **Before the image pull**
+    - An image pull intent record is written befor the actual pull occurs.
+    - This is is to mitigate a situation where the kubelet crashes during the pull and an
+      image that gets pulled in the meantime later appears as if it were *preloaded*.
+2. **After the pull**
+    - Once we know the result of the pull operation, we either remove the image pull
+      intent record (=> failed pull attempt), or we write a new record of the image
+      being pulled successfully and remove the record of the pull intent.
+    - Without this, if kubelet failed during an image pull, an existing record would
+      already grant image access irregardless of this pull operation result for any next
+      pull attempts.
+
+The kubelet uses the following Golang interface to interact with the cache:
+```go
+// ImagePullCredentials wraps credentials that can be used to pull an image.
+// One of the attributes must always be non-nil, and the others must be nil.
+type ImagePullCredentials struct {
+    KubernetesSecret *imagemanager.ImagePullSecret
+    NodeWideCredentials *bool
+    Anonymous *bool
+}
+
+// ImagePullManager keeps the state of images that were pulled and which are
+// currently still being pulled.
+// It should keep an internal state of images currently being pulled by the kubelet
+// in order to determine whether to destroy a "pulling" record should an image
+// pull fail.
+type ImagePullManager interface {
+    // RecordPullIntent records an intent to pull an image.
+    // It returns a number of current pulls requesting the same image.
+    RecordPullIntent(image string) (uint23, error)
+    // RecordImagePulled writes a record of an image being successfully pulled
+    // with ImagePullCredentials.
+    //
+    // If `credentials` is nil, it means the image was pulled successfully, but
+    // a reverification should be required upon next pull request. This can be
+    // useful in cases where the kubelet crashed in the middle of the pull but
+    // the image appears on the node after the kubelet restarts.
+    RecordImagePulled(imageRef string, credentials *ImagePullCredentials) error
```

Comment:
I think we *have* to pass in the same `image string` arg we pass to `RecordPullIntent` so refcounting can get cleaned up properly, *and* so the pulled record can partition these credentials by the image that was being pulled (dropping any tag and digest from the image spec).

We *also* need the canonical `imageRef` string returned from CRI.

```go
// image corresponds to the pod spec and the image passed to RecordPullIntent.
// imageRef is the canonical pulled/resolved id returned from the CRI.
RecordImagePulled(image string, imageRef string, credentials *ImagePullCredentials) error
```